### PR TITLE
Implement job scale up and down

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
Address #782 

Note: 
1. You are responsible to watch the hostfile changes under `/etc/volcano` if you plan to do scale up/down later.
2. The env `VC_{task name}_HOSTS` and `VC_{task name}_NUM` can not be used when you are using scale up/down.

### CODE REVIEW

File: pkg/controllers/job/job_controller.go [hunk 1]
```diff
@@ -215,6 +215,7 @@ func NewJobController(
 	// Register actions
 	state.SyncJob = cc.syncJob
 	state.KillJob = cc.killJob
+	state.UpdateJob = cc.updateJobFn
```
Comment (k82cn) [inline pkg/controllers/job/job_controller.go:218]: can we handle this in SyncJob?

Comment (hzxuzhonghu) [reply pkg/controllers/job/job_controller.go:218]: Good catch. I have tried and it is tricky. Maybe a separate prin future, there can be lots of flags to control different behavior.

**Comment by hzxuzhonghu** [comment]
Note: there is one leftover issue i can not tackle:
In the svc plugin we do inject envs "VC_%s_HOSTS" and "VC_%s_NUM". They can not be changed without restart pod.  
What's lucky is that: I have not heard any user adopted them

File: pkg/controllers/job/job_controller.go (hunk 1 above)
Comment (k82cn) [reply pkg/controllers/job/job_controller.go:218]: similar to statefulset controller, in syncJob, it just make sure pod & replicas keep sync.

**Comment by k82cn** [comment]
> Note: there is one leftover issue i can not tackle:
> In the svc plugin we do inject envs "VC_%s_HOSTS" and "VC_%s_NUM". They can not be changed without restart pod.
> What's lucky is that: I have not heard any user adopted them

Update this limitation into design doc :)

**Comment by hzxuzhonghu** [comment]
>Update this limitation into design doc :)

How about delete these two envs? IMO, it is a little hacky to do so and error prone.

**Comment by k82cn** [comment]
> > Update this limitation into design doc :)
> 
> How about delete these two envs? IMO, it is a little hacky to do so and error prone.

That's ok to me; please also highlight in the document :)

File: pkg/controllers/job/job_controller.go (hunk 1 above)
Comment (k82cn) [reply pkg/controllers/job/job_controller.go:218]: any update?

File: pkg/controllers/job/job_controller_actions.go [hunk 2]
```diff
@@ -300,7 +327,11 @@ func (cc *Controller) syncJob(jobInfo *apis.JobInfo, updateStatus state.UpdateSt
 		return fmt.Errorf("failed to create %d pods of %d", len(creationErrs), len(podToCreate))
 	}
 
-	// TODO: Can hardly imagine when this is necessary.
+	// Delete pods when scale down.
+	// sort pods to be deleted from higher order to lower
+	sort.Slice(podToDelete, func(i, j int) bool {
+		return podToDelete[i].Name > podToDelete[j].Name
```
Comment (k82cn) [inline pkg/controllers/job/job_controller_actions.go:333]: how about 1 vs 12

**Comment by hzxuzhonghu** [comment]
@k82cn take another round of review

File: pkg/controllers/job/job_controller_actions.go [hunk 3]
```diff
@@ -300,7 +329,15 @@ func (cc *Controller) syncJob(jobInfo *apis.JobInfo, updateStatus state.UpdateSt
 		return fmt.Errorf("failed to create %d pods of %d", len(creationErrs), len(podToCreate))
 	}
 
-	// TODO: Can hardly imagine when this is necessary.
+	// Delete pods when scale down.
+	// sort pods to be deleted from higher order to lower
+	sort.Slice(podToDelete, func(i, j int) bool {
+		iName := strings.Split(podToDelete[i].Name, "-")
```
Comment (k82cn) [inline pkg/controllers/job/job_controller_actions.go:335]: can we use "volcano.sh/task-index" annotation?

Comment (k82cn) [reply pkg/controllers/job/job_controller_actions.go:335]: as we forbide volcano.sh annotation update in admission controller.

**Pull request merged**